sklearn/ensemble/_forest.py

```diff
@@ -398,13 +408,63 @@ def fit(self, X, y, sample_weight=None):
         if self.oob_score:
             self._set_oob_score(X, y)
 
+        if self.feature_importances_type == "permutation":
+            self._set_oob_permutation_importance(X, y)
+
         # Decapsulate classes_ attributes
         if hasattr(self, "classes_") and self.n_outputs_ == 1:
             self.n_classes_ = self.n_classes_[0]
             self.classes_ = self.classes_[0]
 
         return self
 
+    def _set_oob_permutation_importance(self, X, y):
+        """
+        Calculate out of bag predictions and score."""
+        X = check_array(X, dtype=DTYPE, accept_sparse=False)
+
+        n_samples = y.shape[0]
+
+        n_predictions = np.zeros(n_samples)
+
+        n_samples_bootstrap = _get_n_samples_bootstrap(
+            n_samples, self.max_samples
+        )
+
+        all_imp = np.zeros((self.n_estimators, X.shape[1]))
+        for i, estimator in enumerate(self.estimators_):
+            random_state = check_random_state(estimator.random_state)
+
+            unsampled_indices = _generate_unsampled_indices(
+                estimator.random_state, n_samples, n_samples_bootstrap
+            )
+
+            scorer = check_scoring(estimator, scoring=self.scoring)
+            baseline = scorer(
+                estimator, X[unsampled_indices, :], y[unsampled_indices, :]
+            )
+
+            for col in range(X.shape[1]):
+                X_permuted = X.copy()
```

Comment:
And I am almost sure that we could benefit from the `sklearn.inspection._permutation_importance._calculate_permutation_scores` function that given all input will compute the score.


We can also parallelize one of the `for` loop. I would imagine that it would be best to parallelize the external loop, meaning on the number of estimators.

In the other permutation importance scheme, we do multiple rounds. I assume that here we don't need to do that since we rely on the bag of estimators to have an estimate of the variance as well.